applications/ChatGPT/chatgpt/trainer/ppo.py

```diff
@@ -75,25 +76,32 @@ def __init__(self,
         self.critic_optim = critic_optim
         self._set_default_generate_kwargs(generate_kwargs, actor)
 
+        self.kl_coef = kl_coef
+
     def training_step(self, experience: Experience) -> Dict[str, float]:
         self.actor.train()
         self.critic.train()
 
         num_actions = experience.action_mask.size(1)
         action_log_probs = self.actor(experience.sequences, num_actions, attention_mask=experience.attention_mask)
+
+        reward_with_penalty = compute_reward(experience.reward, self.kl_coef, action_log_probs, experience.base_action_log_probs, action_mask=action_mask)
+        advantage_with_penalty = reward_with_penalty - experience.values
         actor_loss = self.actor_loss_fn(action_log_probs,
                                         experience.action_log_probs,
-                                        experience.advantages,
+                                        advantage_with_penalty,
                                         action_mask=experience.action_mask)
         self.strategy.backward(actor_loss, self.actor, self.actor_optim)
         self.strategy.optimizer_step(self.actor_optim)
         self.actor_optim.zero_grad()
 
+        values = self.critic(experience.sequences,
+                             action_mask=experience.action_mask,
+                             attention_mask=experience.attention_mask)
```

Comment:
Why to calculate values for 2 times here?

Reply:
redundant, one should be removed